src/Algebra/Graph/Bipartite/AdjacencyMap.hs

```diff
@@ -650,6 +662,141 @@ vertexSet = Set.fromAscList . vertexList
 edgeSet :: (Ord a, Ord b) => AdjacencyMap a b -> Set.Set (a, b)
 edgeSet = Set.fromAscList . edgeList
 
+-- | The /circuit/ on a list of vertices.
+-- Complexity: /O(n * log(n))/ time and /O(n)/ memory.
+--
+-- @
+-- circuit []                       == 'empty'
+-- circuit [(x, y)]                 == 'edge' x y
+-- circuit [(x, y), (z, w)]         == 'connect' ('vertices' [x, z] []) ('vertices' [] [y, w])
+-- circuit [(1, 2), (3, 4), (5, 6)] == swap 1 * (2 + 6) + swap 3 * (2 + 4) + swap 5 * (6 + 2)
+-- circuit . 'reverse'              == 'swap' . circuit . 'map' 'Data.Tuple.swap'
```

Comment:
That's a mind-bending property, I had to draw a picture to see what's going on. Nice :)